modules/contracts/src.sol/CMCWithdraw.sol

```diff
@@ -3,57 +3,98 @@ pragma solidity ^0.7.1;
 pragma experimental ABIEncoderV2;
 
 import "./interfaces/ICMCWithdraw.sol";
+import "./interfaces/Types.sol";
+import "./interfaces/WithdrawHelper.sol";
 import "./CMCCore.sol";
 import "./AssetTransfer.sol";
 import "./lib/LibAsset.sol";
 import "./lib/LibChannelCrypto.sol";
 import "./lib/LibUtils.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
 
 contract CMCWithdraw is CMCCore, AssetTransfer, ICMCWithdraw {
   using LibChannelCrypto for bytes32;
 
   mapping(bytes32 => bool) private isExecuted;
 
-  /// @param recipient The address to which we're withdrawing funds to
-  /// @param assetId The token address of the asset we're withdrawing (address(0)=eth)
-  /// @param maxAmount The number of units of asset we're withdrawing
+  modifier validateWithdrawData(WithdrawData calldata wd) {
+    require(
+      wd.channelAddress == address(this),
+      "CMCWithdraw: Channel address mismatch"
+    );
+    _;
+  }
+
+  /// @param wd The withdraw data consisting of
+  /// semantic withdraw information, i.e. assetId, recipient, and amount;
+  /// information to make an optional call in addition to the actual transfer,
+  /// i.e. target address for the call and call payload;
+  /// additional information, i.e. channel address and nonce.
   /// @param aliceSignature Signature of owner a
   /// @param bobSignature Signature of owner b
   function withdraw(
-    address payable recipient,
-    address assetId,
-    uint256 maxAmount,
-    uint256 nonce,
-    bytes memory aliceSignature,
-    bytes memory bobSignature
-  ) external override onlyViaProxy nonReentrant {
-    // Replay protection
-    bytes32 withdrawHash = keccak256(abi.encodePacked(recipient, assetId, maxAmount, nonce));
-    require(!isExecuted[withdrawHash], "CMCWithdraw: Transaction has already been executed");
-    isExecuted[withdrawHash] = true;
+    WithdrawData calldata wd,
+    bytes calldata aliceSignature,
+    bytes calldata bobSignature
+  )
+    external
+    override
+    onlyViaProxy
+    nonReentrant
+    validateWithdrawData(wd)
+  {
+    // Generate hash
+    bytes32 wdHash = hashWithdrawData(wd);
 
-    // Validate signatures
-    require(withdrawHash.checkSignature(aliceSignature, alice), "CMCWithdraw: Invalid alice signature");
-    require(withdrawHash.checkSignature(bobSignature, bob), "CMCWithdraw: Invalid bob signature");
+    // Verify Alice's and Bob's signature on the withdraw data
+    verifySignatures(wdHash, aliceSignature, bobSignature);
+
+    // Replay protection
+    require(!isExecuted[wdHash], "CMCWithdraw: Transaction has already been executed");
+    isExecuted[wdHash] = true;
 
     // Determine actually transferable amount
-    uint256 balance = LibAsset.getOwnBalance(assetId);
-    uint256 amount = LibUtils.min(maxAmount, balance);
+    uint256 balance = LibAsset.getOwnBalance(wd.assetId);
+    uint256 amount = LibUtils.min(wd.amount, balance);
+
+    // TODO: Should we revert if amount is 0?
```

Comment:
There could be some cases where we want to execute the additional call, even with a 0 balance (can't think of any, but I'm sure someone else could) so we probably don't want to revert here unless the `amount` and `callTo` are empty

Reply:
I agree. In fact, I don't even think it's necessary to revert if both `amount` and `callTo` are zero; we could just "let it run" in this case too (which is what I implemented, but I felt I shouldn't "silently" decide this on my own, so I added the comment).
Do you think reverting would be better if both are zero? (One thing that might be worth considering in this context are retries of failed transactions.)

Reply:
We only retry on known errors (i.e. nonce errors), so I don't think we have to worry too much about the retries on reverts.

I think reverting iff both are zero is a good call, since its very likely that's not what you wanted to do anyways